### PR TITLE
Update facts to break down to individual properties

### PR DESCRIPTION
**Pull request by mundruid**
- Broke down facts to individual properties in base device and all devices that had implemented facts
- Modified unit tests for properties
- Addresses issue #99

### CODE REVIEW

File: pyntc/devices/eos_device.py [hunk 1]
```diff
@@ -187,25 +191,81 @@ def enable(self):
...
+        return self._uptime_string
+
+    @property
+    def hostname(self):
+        if self._hostname is None:
             sh_hostname_output = self.show("show hostname")
-            self._facts.update(
-                convert_dict_by_key(sh_hostname_output, {}, fill_in=True, whitelist=["hostname", "fqdn"])
-            )
+            self._hostname = convert_dict_by_key(sh_hostname_output, {}, fill_in=True, whitelist=["hostname"])[
```
Comment (jmcgill298) [inline pyntc/devices/eos_device.py:218]: I believe we can just do `self._hostname = show_hostname_output["hostname"]`

File: pyntc/devices/eos_device.py [hunk 2]
```diff
@@ -187,25 +191,81 @@ def enable(self):
...
+        if self._vlans is None:
+            self._vlans = self._get_vlan_list()
+
+        return self._vlans
+
+    @property
+    def fqdn(self):
+        if self._fqdn is None:
+            sh_hostname_output = self.show("show hostname")
+            self._fqdn = convert_dict_by_key(sh_hostname_output, {}, fill_in=True, whitelist=["fqdn"])["fqdn"]
```
Comment (jmcgill298) [inline pyntc/devices/eos_device.py:242]: I believe we can just do `self._fqdn= show_hostname_output["fqdn"]`

File: pyntc/devices/eos_device.py [hunk 3]
```diff
@@ -187,25 +191,81 @@ def enable(self):
...
+            ]
+
+        return self._hostname
 
-            self._facts["interfaces"] = self._get_interface_list()
-            self._facts["vlans"] = self._get_vlan_list()
+    @property
+    def interfaces(self):
+        if self._interfaces is None:
+            self._interfaces = self._get_interface_list()
```
Comment (jmcgill298) [inline pyntc/devices/eos_device.py:227]: lets move this logic from the internal method to hear and whatever needs that can just reference the property

File: pyntc/devices/eos_device.py [hunk 4]
```diff
@@ -187,25 +191,81 @@ def enable(self):
...
+
+        return self._interfaces
+
+    @property
+    def vlans(self):
+        if self._vlans is None:
+            self._vlans = self._get_vlan_list()
```
Comment (jmcgill298) [inline pyntc/devices/eos_device.py:234]: lets move this logic from the internal method to hear and whatever needs that can just reference the property

File: pyntc/devices/ios_device.py [hunk 5]
```diff
@@ -306,27 +305,93 @@ def enable(self):
...
+        if self._vendor is None:
+            self._vendor = "cisco"
+
+        return self._vendor
+
+    @property
+    def uptime(self):
+        version_data = self._raw_version_data()
+        uptime_full_string = version_data["uptime"]
+        if self._uptime is None:
```
Comment (jmcgill298) [inline pyntc/devices/ios_device.py:318]: This should be moved to the first line of the property

File: pyntc/devices/ios_device.py [hunk 6]
```diff
@@ -306,27 +305,93 @@ def enable(self):
...
+            self._uptime = self._uptime_to_seconds(uptime_full_string)
+
+        return self._uptime
+
+    @property
+    def uptime_string(self):
+        version_data = self._raw_version_data()
+        uptime_full_string = version_data["uptime"]
+        if self._uptime_string is None:
```
Comment (jmcgill298) [inline pyntc/devices/ios_device.py:327]: This should be moved to the first line of the property

File: pyntc/devices/ios_device.py [hunk 7]
```diff
@@ -306,27 +305,93 @@ def enable(self):
...
+    @property
+    def serial_number(self):
+        version_data = self._raw_version_data()
+        if self._serial_number is None:
+            self._serial_number = version_data["serial"]
+
+        return self._serial_number
+
+    @property
+    def cisco_ios_ssh(self):
```
Comment (jmcgill298) [inline pyntc/devices/ios_device.py:389]: lets just change this to a `config_register` property and lose the dictionary

**Review by jmcgill298 — CHANGES_REQUESTED** [review]
Good work @mundruid there are a few comments I have provided, but there are 3 items that should be looked at all around that were not commented each time:

1. Lets keep `vendor` as a class attribute as this should be constant
2. The `_get_*` methods were created to simplify the large `facts` property. Since this is being broken into smaller individual pieces, the logic in those internal methods can be moved to the respective property
3. The `if` conditional for checking if the property has already been discovered should be at the top of each property. This will speed things up and prevent wasting time/cpu resources trying to run commands on the device


Thanks

**Comment by mundruid** [comment]
> Good work @mundruid there are a few comments I have provided, but there are 3 items that should be looked at all around that were not commented each time:
> 
> 1. Lets keep `vendor` as a class attribute as this should be constant
> 2. The `_get_*` methods were created to simplify the large `facts` property. Since this is being broken into smaller individual pieces, the logic in those internal methods can be moved to the respective property
> 3. The `if` conditional for checking if the property has already been discovered should be at the top of each property. This will speed things up and prevent wasting time/cpu resources trying to run commands on the device
> 
> Thanks

@jmcgill298, 

Thanks for the review. These are good points, especially the second one on improving speed. I will address these  by EOBD today so that we can merge since there are several other PRs in the stack.

Thanks!

File: pyntc/devices/base_device.py [hunk 8]
```diff
@@ -30,7 +30,16 @@ def __init__(self, host, username, password, device_type=None, **kwargs):
         self.username = username
         self.password = password
         self.device_type = device_type
-        self._facts = None
+        self._uptime = None
+        self._vendor = None
```
[the remark this answers is not in the document]
Comment (mundruid) [reply pyntc/devices/base_device.py:34]: Addressed in next commit.

File: pyntc/devices/eos_device.py [hunk 9]
```diff
@@ -144,7 +142,13 @@ def _wait_for_device_reboot(self, timeout=3600):
             except:  # noqa E722
                 pass
 
-        raise RebootTimeoutError(hostname=self.facts["hostname"], wait_time=timeout)
+        raise RebootTimeoutError(hostname=self.hostname, wait_time=timeout)
+
+    def _get_uptime(self):
```
[the remark this answers is not in the document]
Comment (mundruid) [reply pyntc/devices/eos_device.py:147]: Addressed in next commit.

**Comment by mundruid** [comment]
@jmcgill298 I have addressed all your comments. Please let me know if I am missing anything.

**Review by jmcgill298 — APPROVED** [review]

**Comment by jmcgill298** [comment]
Thanks @mundruid

**Pull request merged**